test/message-list.test.js

```diff
@@ -0,0 +1,89 @@
+import { expect } from '@esm-bundle/chai';
+import { fixtureSync } from '@open-wc/testing-helpers';
+import { afterNextRender } from '@polymer/polymer/lib/utils/render-status.js';
+import '../vaadin-message-list.js';
+
+function nextRender(target) {
+  return new Promise((resolve) => {
+    afterNextRender(target, () => {
+      resolve();
+    });
+  });
+}
+
+describe('message-list', () => {
+  let messageList, messages;
+
+  beforeEach(() => {
+    let root = document.documentElement;
+    root.style.setProperty('--vaadin-user-color-1', 'purple');
+    root.style.setProperty('--vaadin-user-color-2', 'blue');
+
+    messageList = fixtureSync('<vaadin-message-list></vaadin-message-list>');
+    messages = [
+      {
+        text: 'A message in the stream of messages',
+        time: '9:34 AM',
+        user: {
+          name: 'Joan Doe',
+          abbr: 'JD',
+          img: '/test/visual/avatars/avatar.jpg',
+          colorIndex: 1
+        }
+      },
+      {
+        text: 'Call upon the times of glory',
+        time: '2:34 PM',
+        user: {
+          name: 'Steve Mops',
+          abbr: 'SM',
+          colorIndex: 2
+        }
+      }
+    ];
+  });
+
+  it('should have a valid version number', () => {
+    expect(messageList.constructor.version).to.match(/^(\d+\.)?(\d+\.)?(\d+)(-(alpha|beta)\d+)?$/);
+  });
+
+  it('message list should be initially empty', () => {
+    expect(messageList.items).to.be.empty;
+  });
+
+  describe('items property', () => {
+    beforeEach(async () => {
+      messageList.items = messages;
+      await nextRender(messageList);
+    });
+
+    it('message list should have two messages', () => {
+      const items = messageList.shadowRoot.querySelectorAll('vaadin-message');
+      expect(items.length).to.equal(2);
+    });
+
+    it('message properties should be correctly set', () => {
+      const firstMessage = messageList.shadowRoot.querySelectorAll('vaadin-message')[0];
+      const content = firstMessage.shadowRoot.querySelector('[part="content"]');
+      const time = firstMessage.shadowRoot.querySelector('[part="time"]');
+      const name = firstMessage.shadowRoot.querySelector('[part="name"]');
+      const avatar = firstMessage.shadowRoot.querySelector('vaadin-avatar');
+
+      // Test that the <vaadin-message> slotted content is correct
+      const slot = content.querySelector('slot');
+      const nodes = slot.assignedNodes({ flatten: true });
+      expect(nodes[0].textContent).to.be.equal(messages[0].text);
+
+      // Test the direct dependencies
+      expect(time.textContent).to.be.equal(messages[0].time);
+      expect(name.textContent).to.be.equal(messages[0].user.name);
```

Comment:
These checks also rely on the implementation details of `vaadin-message`. 
Consider replacing them with another assertion that checks if the `user` property is set.

Is the test written this way to avoid updating it after #7 is implemented?
If so, it would be preferable to update the test later on when changing API.

Reply:
This was written before I new about #7, so it has not been taken into account.

I disagree a bit with testing the API only. If the API changes, e.g. If `user` goes away and you should user `userName`, then the tests will still pass because `user` is indeed set. It is not just used for anything.

How should I write the test so that it doesn't rely on internals, but still verify that the value that I give it is actually used?

Reply:
This is indeed a problem which should be however covered with a different kind of test (e.g. visual test).
I assume it fails when we make this type of change in `vaadin-message` but not in `vaadin-message-list`.

Feel free to discard my review if you think the test makes more sense as it is.

Reply:
My opinion is going back and forth along your discussion. :D
- Yes, it's bad to depend on implementation details of `vaadin-message`.
- But because of lacking type-safety, the test indeed doesn't have much value if it just checks setting some property (which might or might not do something) on an element.
- But then again it's a good point that this problem would be caught by visual tests.

But then it's questionable whether we need these unit tests at all, if we can't trust them, and depend on the visual tests instead.

I'm fine either way. Maybe I would prioritize safety and recommend to keep them _the Jens way_.

Reply:
I still think it would be good to see in unit tests that we are setting the correct properties of message when configuring the list. To get forward, however, we have to make some call. I went with the opinion of our most seasoned expert on our components. It will also ensure more consistency across the component set. 

Now the test only checks that avatar.user and avatar.test is set, and checks the textContent that the text of the message is set.

If we want to continue on this discussion about verifying in unit tests if we are setting the correct attributes, then let's do it in a separate issue.

Reply:
Done.